### PR TITLE
Allow delegated IPAM to specify uplink interface

### LINKED ISSUE #34604: CFP: add Volcengine cloud ENI IPAM support
**Describe the feature you'd like**

Volcengine(Volcano Engine) is a cloud platform provided by ByteDance. In cloud environment, the form of placing Pods and hosts at the same network layer (VPC network) to get better performance and operability has been widely accepted. The Cilium users on Volcengine platform are calling for the ENI IPAM support just like the mode they are using on AWS and Alibaba Cloud:

- https://github.com/cilium/cilium/pull/8347
- https://github.com/cilium/cilium/pull/15160

So we would like to support the Volcengine cloud ENI IPAM in Cilium.

**(Optional) Describe your proposed solution**

Support for Volcengine cloud ENI IPAM will be split into several sub-tasks:

- [ ] Add Volcengine ENI IPAM fields into CiliumNode CRD. #34608

- [ ] Add Volcengine IPAM provider in Cilium Operator.

- [ ] Setup routing for Pod using Volcengine ENI in dataplane.

- [ ] Add Volcengine IPAM docs

### PR DESCRIPTION
**Pull request by ruicao93**
For the scenario where there are multiple interfaces on a Node and delegated IPAM is used to determine the IP and the uplink interface of a Pod.  The "interfaces" field can be used by IPAM plugin to pass uplink interface info to Cilium. And then Cilium can setup host routes for the Pod according to the uplink interface info.

The "interfaces" field is allowed to specify the related host interface according to the CNI spec and is already widely used by existing plugins, such as bridge and ptp plugin. It's actually not used by any IPAM plugin for now, but I think it's reasonable and conforms to the original definition of the field.

Motivations/Benefits of the PR for the native routing mode:

- More cloud(public or private) envs with multiple uplink can be easily ingregrated with Cilium without involving any cloud specific logic. A common way is provided for the case: "uplink interface" support.
- Cilium keeps full control of the Pod traffic. Cilium choose the implement of the "routing" for egress traffic, through Linux routes or eBPF forwarding rules or any other methods.

[- `interfaces`: An array of all interfaces created by the attachment, including any host-level interfaces:
](https://github.com/containernetworking/cni/blob/4c9ae43c0eaa85ec1ab27781e9b258f13e7fd0ca/SPEC.md?plain=1#L571)

[bridge.go
](https://github.com/containernetworking/plugins/blob/06ba001d846042799378f4cb76e085909bbf4723/plugins/main/bridge/bridge.go#L563)
``` go
	// Assume L2 interface only
	result := &current.Result{
		CNIVersion: current.ImplementedSpecVersion,
		Interfaces: []*current.Interface{
			brInterface,
			hostInterface,
			containerInterface,
		},
	}
```

bridge plugin result:

```json
{
    "ips": [
        {
          "address": "10.1.0.5/16",
          "gateway": "10.1.0.1",
          "interface": 2
        }
    ],
    "routes": [
      {
        "dst": "0.0.0.0/0"
      }
    ],
    "interfaces": [
        {
            "name": "cni0",
            "mac": "00:11:22:33:44:55"
        },
        {
            "name": "veth3243",
            "mac": "55:44:33:22:11:11"
        },
        {
            "name": "eth0",
            "mac": "99:88:77:66:55:44",
            "sandbox": "/var/run/netns/blue"
        }
    ],
    "dns": {
      "nameservers": [ "10.1.0.1" ]
    }
}
```

Fixes: #34604

### CODE REVIEW

**Comment by joamaki** [comment]
As the title still has `Draft` in it, I'll wait before reviewing. Edit the title and ping me when you think it's ready.

**Comment by ruicao93** [comment]
> As the title still has `Draft` in it, I'll wait before reviewing. Edit the title and ping me when you think it's ready.

Thanks @joamaki, the commit is ready for review and more discussion could be found in https://github.com/cilium/cilium/issues/34604

**Comment by ruicao93** [comment]
Hi @joestringer @gandro , could you please help review the PR as we discussed.
@cilium/sig-ipam

**Comment by joestringer** [comment]
I'm not so familiar with delegated IPAM, but one crucial part of this will be to test that this doesn't break existing delegated IPAM setups. I see that issue https://github.com/cilium/cilium/issues/34788 is filed to create automated testing for that.

**Comment by joestringer** [comment]
I see that there is now https://github.com/cilium/cilium/pull/34839 out for review, so it should be possible to combine that test with this PR and test it out on your own repository?

**Comment by joestringer** [comment]
@dylandreimerink I don't think it's meaningful to run `/test` on this PR, as the delegated IPAM path isn't enabled for any of the existing tests merged into the tree.

**Comment by ruicao93** [comment]
> I see that there is now #34839 out for review, so it should be possible to combine that test with this PR and test it out on your own repository?

Sure, l will run the test provided in #34839 .  The Cilium e2e and k8s conformance tests have been verified on Volcengine cloud combined with Volcengine delegated IPAM:

- CIlium e2e tests:

With one minor error, but it doesn't matters.
image

- Kubernetes conformance network tests
Plugin: e2e
Status: passed
Total: 6965
Passed: 42
Failed: 0
Skipped: 6923

File: plugins/cilium-cni/cmd/cmd.go
Comment (gandro) [inline plugins/cilium-cni/cmd/cmd.go]: Why isn't the delegated plugin doing this? I guess we've tested this that this doesn't break the delegated Azure plugin, but I find it very weird that we are installing a rule on behalf of a delegated plugin, if the delegated plugin does have all the information needed to do it itself.

Or am I missing something and there is information here that the delegated plugin does not have? The only thing that comes to mind is the masquerading exception, but that code here does not populate the necessary exclusion CIDR, so that doesn't seem to be it either.

**Review by gandro — COMMENTED** [review]
Thanks! I have a few questions on this approach that are still unclear to me

File: plugins/cilium-cni/cmd/cmd.go
Comment (ruicao93) [reply plugins/cilium-cni/cmd/cmd.go]: @gandro , more contexts could be foun in https://github.com/cilium/cilium/issues/34604 

Routes and masq exception are controlled and should only be controlled by Cilium(main CNI) as following reasons:

- Cilium knowns all routes and masq info. It's better to controlled them in one component. 
- Pod virtual interfaces are yet not created when calling delegated plugin, so ingress routes(control traffic to Pod) has no chance to be installed.

For masquerading cidr, it's still controlled through Cilium "ipv4NativeRoutingCIDR' option.

Comment (gandro) [reply plugins/cilium-cni/cmd/cmd.go]: > Pod virtual interfaces are yet not created when calling delegated plugin, so ingress routes(control traffic to Pod) has no chance to be installed.

I'm not sure I follow. The routes installed by the call to `interfaceAdd` do not contain any interface name. Why do the interfaces need to exist?

Comment (gandro) [reply plugins/cilium-cni/cmd/cmd.go]: Ah. I guess you are talking about the per-endpoint routes (ingress). But those are not installed here, here we are talking about the egress routes. Those are not configured by the CNI, so I don't think this is relevant here.

`interfaceAdd` creates egress rules, steering traffic from a certain source IP over a certain host interface. I strongly believe this is something that the delegated plugin should configure itself, rather than relying on our EC2 logic (which we might also want to change if EC2 changes, which then would break all delegated plugins. I guess that is my main worry)

Comment (ruicao93) [reply plugins/cilium-cni/cmd/cmd.go]: @gandro you are right, interfaceAdd() is called to install egress routes, veth/netkit interface number is not needed when installing egress routes.

The main idea is:
1. Control all routes only in Cilium is better than spliting part of them to another plugin(Delegated IPAM).
2. IPAM plugin is designed for IP allocation purpose. Not suitable for modifying system network configuration.
3. Cilium is actually already support multiple interfaces native routing. Which is mainly used in ENI mode. So we could leverage the existing capability, and just leave the IP allocation part to delegated IPAM plugin.

Comment (gandro) [reply plugins/cilium-cni/cmd/cmd.go]: I understand the desire to offload functionality to Cilium, but this adds some tight coupling between Cilium and the delegated IPAM plugin on a very very leaky abstraction boundary:

This here assumes that every delegated plugin that returns a host-level "Interfaces" object will want an Amazon EC2-ENI-style egress rule set up by Cilium. This might be true for Volcegine, but it's not true for example by the delegated Azure IPAM plugin. The Azure IPAM plugin does not seem to return an "Interfaces" object right now, but what if it ever does? Cilium would start to install egress EC2-style rules that would break the Azure plugin, because it doesn't work like that.

> IPAM plugin is designed for IP allocation purpose. Not suitable for modifying system network configuration.

I kindly disagree with this. Any cloud-provider based IPAM implementation will always require system network configuration changes, to teach the underlying fabric on how pod IPs are attached. Some component has to implement it. You are arguing that the plugin is the wrong place. I'm arguing that doing this inside Cilium is the wrong place, because it requires Cilium to understand what a certain delegated IPAM plugin requires and hard-code that cloud-providers specific network configuration. 

The whole idea behind delegated IPAM plugins is that Cilium does no longer have to implement cloud-provider specific logic. Yes, there are the existing ENI IPAM modes that do implement cloud-provider specific logic, but this comes at a high maintenance cost that we want to reduce. This is why we are pushing to move as much logic as we can out of Cilium and into the delegated plugin.

---

Having said that, while I strongly disagree with this approach here where the Cilium CNI plugin interprets the delegated IPAM plugin result as "this plugin wants EC2-style egress routes") - if there is Cilium-specific logic that you need to build upon to implement the delegated plugin (e.g. obtaining the masquerading exceptions for the egress rules), then I think we are open to add endponts to the [Cilium API](https://docs.cilium.io/en/stable/api/#api-ref) to support such use-cases. That makes it explicit what the delegated plugin wants, rather than assuming something.

In general, I think it might be worthwhile to take a step back and have a real CFP document that presents the whole architecture you have in mind for the Volcengine support. For example: 
- What will the delegated IPAM plugin do?
- What Cilium-adjacent CRDs do you expect to implement?
- What functionality (e.g. veth pair setup, ingress/egress routing) do you require of Cilium? How do you plan to invoke this functionality?

That way we can have a better shared understanding of what you expect of Cilium and we can see what kind of functionality we think Cilium can do, and what functionality we think should _not_ be handled by Cilium.

Comment (ruicao93) [reply plugins/cilium-cni/cmd/cmd.go]: >  EC2 logic (which we might also want to change if EC2 changes, which then would break all delegated plugins. I guess that is my main worry)

It's indeed a problem. Actually I believe the current ENI logic dose meet the requirements of most ENI or other multi interface cases. So as an optional way, could we add common multi interface logic if there are concern about breaking  current EC2 logic?

The common multi interface logic could be shared by newly joined cloud platforms or on prime deployments. Let Cilium handle the native routing for this kind of common case. If the 3rd delegation IPAM plugin whish to customize the routes by themself, they cloud not return the uplink hints in the result.

Comment (gandro) [reply plugins/cilium-cni/cmd/cmd.go]: > If the 3rd delegation IPAM plugin whish to customize the routes by themself, they cloud not return the uplink hints in the result.

As mentioned above, I'd prefer it the other way around: Let's have the delegated IPAM plugin call Cilium API if it wants things to be done by Cilium. That way, we can also provide as many Cilium-specific fields as we want, rather than relying information via the CNI result which was not designed for this purpose

Comment (dakehero) [reply plugins/cilium-cni/cmd/cmd.go]: > Let's have the delegated IPAM plugin call Cilium API if it wants things to be done by Cilium

I don't think it is a good way to do such things, it would only bring even more coupling between Cilium and the delegated IPAM plugin. 

As @ruicao93 mentioned above, the delegated IPAM plugin should only do IP allocation and tell main plugin what is the IP address of Pod and to which interface the pod supposed to bind as uplink, and don't care about in which way will main plugin realize traffic steering(using policy routing or bpf routing or event creating an ipvlan slave for pod).

**Comment by ruicao93** [comment]
hi @tamilmani1989. We're discussing whether it is better to install node routes for Pod in Cilium or in delegated IPAM plugin.

I have a few of questions about the Delegated Azure IPAM mode. Could you please help me with them?

- Does the azure ipam will create multiple network interfaces for Pods like how aws ENI mode works?
- If there're multile ENIs, is the azure IPAM compoment responsible for install egress routes for Pods to redirect traffic to right network interface?

File: plugins/cilium-cni/cmd/cmd.go
Comment (gandro) [reply plugins/cilium-cni/cmd/cmd.go]: > I don't think it is a good way to do such things, it would only bring even more coupling between Cilium and the delegated IPAM plugin.

I'd argue the coupling is exactly the same. In both cases Cilium reacts to something that the delegated IPAM plugin requested. The only difference is that the coupling is made explicit via an API designed for that purpose, rather than assuming that when a random delegated plugin returns host interfaces that wants a particular set of egress routing rules to be installed.

Comment (christarazi) [reply plugins/cilium-cni/cmd/cmd.go]: I agree with @gandro's stance here. As the codeowners for the IPAM/cloud code, we are interested in reducing the amount of cloud-specific knowledge required to maintain the code because core Cilium developers don't necessarily have cloud-specific knowledge. While there are cloud IPAM implementations in the code today like ENI, it does not mean we want to open up Cilium for having all cloud-specific logic. Instead, it makes more sense to shift that maintenance to those with the expertise of whichever cloud via a delegated plugin. That plugin would reach out to Cilium API in order to retrieve whatever information it needs to carry out its job. It's an IPAM plugin for Cilium. This keeps the Cilium codebase simpler and less error-prone, while the vast majority of the complexity of the cloud/platform is kept contained inside the plugin. 

We are interested in collaborating with you on a [CFP](https://github.com/cilium/design-cfps/) describing a delegated plugin approach. It might be too soon to be discussing an implementation at this stage.

Comment (dakehero) [reply plugins/cilium-cni/cmd/cmd.go]: > when a random delegated plugin returns host interfaces that wants a particular set of egress routing rules to be installed.

My point is that routing is something ralated to the specific datapath which IPAM plugin is not supposed to be involved into. By storing uplink's information in CNI results, IPAM plugin can give cilium full control of datapath. Cilium can decide how to steering egress traffic to the uplink, in this case native policy routing is used, but cilium can also use bpf to accelerate routing.

**Comment by wedaly** [comment]
> hi @tamilmani1989. We're discussing whether it is better to install node routes for Pod in Cilium or in delegated IPAM plugin.
> 
> I have a few of questions about the Delegated Azure IPAM mode. Could you please help me with them?
> 
> * Does the azure ipam will create multiple network interfaces for Pods like how aws ENI mode works?
> * If there're multile ENIs, is the azure IPAM compoment responsible for install egress routes for Pods to redirect traffic to right network interface?
> 
> Do you think it would be helpful if delegated IPAM plugin only handle IP/interface management and Cilium controls all Pod routes(implemented by linux routes or bpf forwarding rules)?

azure-ipam currently handles only IP address management, not installing routes or multiple network interfaces.

I mentioned this in [the parent issue](https://github.com/cilium/cilium/issues/34604#issuecomment-2338598430), but my understanding is that a conforming delegated IPAM plugin is not allowed to return `interfaces`. (The bridge and ptp plugins are CNI plugins, but are *not* delegated IPAM plugins.) We want azure-ipam to conform to the CNI spec, so I think even if we had this use case we'd look for a different solution or work with the CNI maintainers to extend the spec so the semantics of `interfaces` from a delegated IPAM plugin are well-defined.

**Comment by ruicao93** [comment]
> 
> azure-ipam currently handles only IP address management, not installing routes or multiple network interfaces.
> 
> I mentioned this in [the parent issue](https://github.com/cilium/cilium/issues/34604#issuecomment-2338598430), but my understanding is that a conforming delegated IPAM plugin is not allowed to return `interfaces`. (The bridge and ptp plugins are CNI plugins, but are _not_ delegated IPAM plugins.) We want azure-ipam to conform to the CNI spec, so I think even if we had this use case we'd look for a different solution or work with the CNI maintainers to extend the spec so the semantics of `interfaces` from a delegated IPAM plugin are well-defined.

Thanks @wedaly for your reply. Dose the single uplink(eth0) is enough for AKS? The IPs and bandwidth of single ENI may be limited by the
control plane(quota) or dataplane(performance).

**Comment by squeed** [comment]
Adding @dylandreimerink, who added the "uplink" or "parent" mechanism for ENI.

**Comment by squeed** [comment]
CNI spec author here :-). There's nothing that says the IPAM result is not allowed to contain interfaces; it's not done out of convention. However, if we would like to change the interpretation of the types, I would be amenable to revisiting that in the spec.

The best part is that it wouldn't require any code changes, so we could "ship" it entirely in test.

**Comment by ruicao93** [comment]
> CNI spec author here :-). There's nothing that says the IPAM result is not allowed to contain interfaces; it's not done out of convention. However, if we would like to change the interpretation of the types, I would be amenable to revisiting that in the spec.
> 
> The best part is that it wouldn't require any code changes, so we could "ship" it entirely in text.

Thanks @squeed for the clarification, it's extremely helpful for the current discussion. We also created another PR to update the deletation IPAM result part, could you help review https://github.com/containernetworking/cni/pull/1121 

So it's reasonable to handle the returned uplink hints. Install the routes according to the uplink info could help Cilium have full control of the Pod traffic. Based on the current code modifications, we might also consider adding an option for users to decide whether the routing is controlled by Cilium or controlled by the user.

**Comment by tamilmani1989** [comment]
Late to the party. We have a similar requirement where pod egress traffic to leave via secondary interface instead of primary interface. i noticed eni ipam mode already does this. https://docs.cilium.io/en/latest/network/concepts/routing/#egress. @ruicao93 Does this PR follow the same pattern/code?

File: plugins/cilium-cni/cmd/cmd.go
Comment (tamilmani1989) [reply plugins/cilium-cni/cmd/cmd.go]: i agree on @dakehero point but at same time don't want cilium cni to do custom stuff that fall outside of cni scope. Not sure if CNI/delegated ipam spec supports this @squeed ? if not, we have to standardize. 

@gandro I guess calling cilium api from delegated ipam would increase latency in pod startup time and bring in more complications.

Comment (ruicao93) [reply plugins/cilium-cni/cmd/cmd.go]: @tamilmani1989 it's allowed by CNI spec as squeed said in: https://github.com/cilium/cilium/pull/34779#issuecomment-2414088797

I'm requesting a change to append more description about the uplink in CNI spec:
https://github.com/containernetworking/cni/pull/1121

**Comment by ruicao93** [comment]
> Happy to see we're making progress! The proposal says:
> 
> > * specifying macvlan master
> > * indicating which interface is to be used as the uplink
> 
> How can Cilium distinguish between the two cases? Or to ask a bit more precisely: The upstream spec does not define the semantics of the returned uplink interfaces. The question remains: How do we (Cilium) define the semantics of the upstream interfaces? Do we just statically assume the "ENI-style rule and routes" for anything that the delegated plugin returns? Or do we rely on some other third party information (e.g. an agent flag) on how Cilium should interpret that result?

Yes @gandro, I agree with you. , there is a lack of a clear definition of the purpose of the returned interface. Maybe we could  consider supplementing the purpose of interfaces in the spec.

``` go
type Interface struct {
	Name       string `json:"name"`
	Mac        string `json:"mac,omitempty"`
	Mtu        int    `json:"mtu,omitempty"`
	Sandbox    string `json:"sandbox,omitempty"` // Empty sandbox actually means it's not an interface in Pod.
	+Purpose/Usage string  // Well known purpose could be described in "CNI conventions"
...
}

```

For the second problem. We could add an option in agent to determine if Cilium need to setup ingress/routes for uplink hint by ipam plugin.

**Comment by gandro** [comment]
> For the second problem. We could add an option in agent to determine if Cilium need to setup ingress/routes for uplink hint by ipam plugin.

Yeah, I think worst case we can just assume that anything returned by the delegated plugin has purpose "uplink" and install the ENI rules as long as an agent flag is set to enable this.

**Comment by tamilmani1989** [comment]
> Happy to see we're making progress! The proposal says:
> 
> > * specifying macvlan master
> > * indicating which interface is to be used as the uplink
> 
> How can Cilium distinguish between the two cases? Or to ask a bit more precisely: The upstream spec does not define the semantics of the returned uplink interfaces. The question remains: How do we (Cilium) define the semantics of the upstream interfaces? Do we just statically assume the "ENI-style rule and routes" for anything that the delegated plugin returns? Or do we rely on some other third party information (e.g. an agent flag) on how Cilium should interpret that result?


afaik, It depends on how CNI sets up datapath. For macvlan based cni, it would be interpreted as macvlan master, for bridge based cni, it's bridge interface. For ovs based cni, its ovs switch. Since Cilium does based on L3 routing, it can interpret as uplink interface since thats the master interface

**Comment by dakehero** [comment]
> For the second problem. We could add an option in agent to determine if Cilium needs to setup ingress/routes for uplink hint by ipam plugin.

Apart from that, adding a new field could also be used in chaining mode to hint which host-side interface is uplink for those "meta" plugins like bandwidth. But it's better to do such a change in the next version of the CNI SPEC.

File: plugins/cilium-cni/cmd/cmd.go
Comment (rjdenney) [inline plugins/cilium-cni/cmd/cmd.go:246]: Is there a reason that we are including the gateway for V4 and not for V6?

File: plugins/cilium-cni/cmd/cmd.go
Comment (ecigar13) [inline plugins/cilium-cni/cmd/cmd.go:238]: Why do we go by InterfaceNumber instead of interface name?
If we have eth0 and eth1... or ethn, then going by interface name is more precise, right?

File: plugins/cilium-cni/cmd/cmd.go
Comment (ruicao93) [reply plugins/cilium-cni/cmd/cmd.go:246]: @rjdenney, v6 gateway is included in the latest update.

File: plugins/cilium-cni/cmd/cmd.go
Comment (ruicao93) [reply plugins/cilium-cni/cmd/cmd.go:238]: @ecigar13 `InterfaceNumber` is  used by ENI logic(we also reuse the codes to install routes) to determine the per-ENI route table ID. 

Current patch has not yet manage the interface number allocation for uplinks. So we should set option.Config.EgressMultiHomeIPRuleCompat=true, which will use ifindex as route table ID. We could figure out the iface number mgmt in the next PR.

File: plugins/cilium-cni/cmd/cmd.go
Comment (rjdenney) [inline plugins/cilium-cni/cmd/cmd.go]: What needs to be done to fully support IPV6?

File: plugins/cilium-cni/cmd/cmd.go
Comment (rjdenney) [reply plugins/cilium-cni/cmd/cmd.go:246]: Is there anything else we will need to do to fully support V6?

File: plugins/cilium-cni/cmd/cmd.go
Comment (ruicao93) [reply plugins/cilium-cni/cmd/cmd.go]: @rjdenney The better way is to adjust current ENI logic, add v6 logics along v4 ops. This way Delegated IPAM plugin and ENI IPAM could share same logic. 

Dose azure need v6 for now? Consider the current PR contains too much changes, I will push another one for v6.

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (gandro) [inline install/kubernetes/cilium/values.yaml.tmpl]: IMO this should be `ipam.installUplinkRoutesForDelegatedIPAM`, not `ipam.operator.installUplinkRoutesForDelegatedIPAM`, it's not a setting that applies to cilium-operator after all.

File: plugins/cilium-cni/cmd/cmd.go
Comment (gandro) [inline plugins/cilium-cni/cmd/cmd.go]: I have two points as feedback here:

1. This needs to be `safenetlink.LinkByName()` (see its godoc for details)
2. In the past (and probably even present), we had problems on ENI mode where the cloud provider (i.e. the Amazon EC2 API in our case) would attach a new interface and an IP to the local node (and thus became available in IPAM), but the interface itself was yet discovered by the local Linux kernel. Thus causing such `LinkByName` lookups to fail until the interface was actually attached.

I think it would be worth thinking about how this code deals with such cases (IPAM plugin believes an interface exist, but the local kernel has not seen it yet, because it's still in the process of being attached): Do we return an error from CNI ADD and let kublet retry later (hopefully at a time when the interface will exist), or do we attempt to wait here?

The first one (letting CNI ADD fail) is probably the easier to implement. For that to work, you will have to return an error here and make sure CNI ADD fails, rather than just logging it here.

File: plugins/cilium-cni/cmd/cmd.go
Comment (gandro) [inline plugins/cilium-cni/cmd/cmd.go]: nit: Could we extract a helper function and replace the whole switch statement with an `if needsEndpointRoutingRule(conf) { interfaceAdd(...) }`?

File: plugins/cilium-cni/cmd/cmd.go
Comment (gandro) [reply plugins/cilium-cni/cmd/cmd.go:238]: I'd really like us to get rid of `EgressMultiHomeIPRuleCompat` some day. So I'm glad to hear that you are planning to fix things such that it is not required by patch either.

**Review by gandro — COMMENTED** [review]
Awesome work, thank you for addressing the feedback. I have some minor things, but high-level I think we are on the same page :rocket:

**Comment by gandro** [comment]
Please also note that the branch needs a rebase, as there are conflicts against Helm files

File: plugins/cilium-cni/cmd/cmd.go
Comment (rjdenney) [reply plugins/cilium-cni/cmd/cmd.go]: Yes, for azure we are currently working to get v6 going.

Right now, we are working to get Dualstack VNET Prefix on Secondary NIC. I have done some testing with your changes before you updates yesterday and I could see the routing for the V4 IPs and tested them to make sure they are working but the V6 IPs were not being assigned to the pod or being routed.

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (ruicao93) [reply install/kubernetes/cilium/values.yaml.tmpl]: @gandro yes, I intented to use `ipam.installUplinkRoutesForDelegatedIPAM` but placed to the wrong section. Will update later.

File: plugins/cilium-cni/cmd/cmd.go
Comment (ruicao93) [reply plugins/cilium-cni/cmd/cmd.go]: Thanks @gandro, it should return error if link could not be found yet.

File: plugins/cilium-cni/cmd/cmd.go
Comment (ruicao93) [reply plugins/cilium-cni/cmd/cmd.go]: @rjdenney thanks for your response. Does v4 works well on azure? For v6 and vnet prefix:
- v6: we could support in another PR after this one merged.
- ip prefix: Current logic also works for ip prefix.  But it's indeed no need to install egress rule for every IP. I think there are two possible ways:
    - Return ip prefix of uplink by IPAM. It would need to update CNI spec to add extra retult field(e.g. HostRoutes)
    - (Eazy)Manage uplink egress rules by IPAM itself. I'm not sure whether this complies with the CNI spec.

**Review by gandro — APPROVED** [review]
Great, thank you!

**Review by gandro — COMMENTED** [review]
Please don't forget to squash your commits into one

**Comment by ruicao93** [comment]
> Great, thank you!

Thanks @gandro for your comments! Could you help trigger the CI workflows?

**Comment by gandro** [comment]
Yes, I can do that. But before, could also squash your two commits into one commit? Otherwise the CI results will be invalidated once the commits are changed again.

**Comment by ruicao93** [comment]
> Yes, I can gladly do that. But before, could also squash your two commits into one commit? Otherwise the CI results will be invalidated once the commits are changed again.

Okay, done. I thought just need to squash on merge.

**Review by tklauser — APPROVED** [review]

**Comment by ruicao93** [comment]
Rebsed the commit. @gandro @tklauser , could you help run the ci tests?

**Comment by gandro** [comment]
@ruicao93 There is no need to rebase if CI fails. We can restart individual tests separately (assuming those are flakes). In fact. You should only rebase if a) there are conflicts or b) there are persistent CI failures because the tests have been updated on `main`.

**Comment by ruicao93** [comment]
> @ruicao93 There is no need to rebase if CI fails. We can restart individual tests separately (assuming those are flakes). In fact. You should only rebase if a) there are conflicts or b) there are persistent CI failures because the tests have been updated on `main`.

@gandro Thanks~ Gotcha~

File: plugins/cilium-cni/cmd/cmd.go
Comment (rjdenney) [reply plugins/cilium-cni/cmd/cmd.go]: @ruicao93 V4 does seem to be working well for azure we will just need V6 support. 

And like you mentioned it is working for ip prefix as well and is adding a new rule for each pod. For now this is fine while we test but if you do plan to add either of the two implementations you mentioned let us know.

Also just tagging @tamilmani1989 for visibility.

**Comment by ruicao93** [comment]
> > @ruicao93 There is no need to rebase if CI fails. We can restart individual tests separately (assuming those are flakes). In fact. You should only rebase if a) there are conflicts or b) there are persistent CI failures because the tests have been updated on `main`.
> 
> @gandro Thanks~ Gotcha~

@gandro Finally... all CI cases has passed. Dose the PR could be merged?

**Comment by gandro** [comment]
There is still a review needed from the @cilium/api CODEOWNER - the designated reviewer is @learnitall

File: plugins/cilium-cni/cmd/cmd.go
Comment (rjdenney) [reply plugins/cilium-cni/cmd/cmd.go]: @ruicao93 Also do we have a date that we can expect the V6 changes PR to be out?

**Review by learnitall — APPROVED** [review]
API changes LGTM! Thank you for the ping, I apologize for the delay.

File: plugins/cilium-cni/cmd/cmd.go
Comment (tamilmani1989) [reply plugins/cilium-cni/cmd/cmd.go]: @rjdenney we can take up for v6

**Comment by joestringer** [comment]
@squeed @gandro I take it from the above discussion that with the spec loosening in https://github.com/containernetworking/cni/pull/1137 we are happy with this solution? Just want to check given that this is now approved by all codeowners and the MLH bot has marked the PR as ready-to-merge.

File: plugins/cilium-cni/cmd/cmd.go
Comment (ruicao93) [reply plugins/cilium-cni/cmd/cmd.go]: @rjdenney @tamilmani1989 I plan to start adapting to IPv6 after this PR is merged. You can take the v6 adaption if urgent.

**Comment by gandro** [comment]
> @squeed @gandro I take it from the above discussion that with the spec loosening in [containernetworking/cni#1137](https://github.com/containernetworking/cni/pull/1137) we are happy with this solution? Just want to check given that this is now approved by all codeowners and the MLH bot has marked the PR as ready-to-merge.

Yes, from my side I'm happy with the proposed solution. The functionality in this PR is opt-in, so I see a low risk of Cilium accidentally misinterpreting a delegated plugin's result.

**Comment by rjdenney** [comment]
@ruicao93 Do you have a timeline for the additional V6 changes that would be needed, or could you tell us what would need to be added so that we can try working on it from our side?

**Pull request merged**
